contracts/Marketplace.sol

```diff
@@ -1,74 +1,296 @@
+    uint previousPrice,
+    uint newPrice
+  );
+
+  event ServiceVersionCreated(
+    uint indexed serviceIndex,
+    bytes sid,
+    bytes20 hash,
+    bytes url
+  );
+
+  event ServicePaid(
+    uint indexed serviceIndex,
+    bytes sid,
+    address indexed purchaser,
+    address indexed seller,
+    uint256 price
+  );
+
+  // ------------------------------------------------------
+  // Modifier functions
+  // ------------------------------------------------------
+
+  modifier onlyServiceOwner(uint serviceIndex) {
+    require(isServiceOwner(serviceIndex), "Service owner is not the same as the sender");
+    _;
+  }
+
+  // ------------------------------------------------------
+  // Utils functions
+  // ------------------------------------------------------
 
-    event VersionCreated(bytes32 sid, bytes32 id, bytes32 location, uint256 price);
+  function compareBytes(bytes memory a, bytes memory b) internal pure returns (bool) {
+    return keccak256(a) == keccak256(b);
+  }
 
-    constructor(address creator, bytes32 _sid) Ownable() Pausable() public {
-        sid = _sid;
-        transferOwnership(creator);
-        addPauser(creator);
-        renouncePauser();
+  // ------------------------------------------------------
+  // View functions
+  // ------------------------------------------------------
+
+  function isServiceOwner(uint serviceIndex) public view returns (bool) {
+    return services[serviceIndex].owner == msg.sender;
+  }
+
+  function isServiceSidExist(bytes memory sid) public view returns (bool) {
+    for (uint i = 0; i < services.length; i++) {
+      if (compareBytes(services[i].sid, sid)) {
+        return true;
+      }
     }
+    return false;
+  }
 
-    function createVersion(bytes32 versionHash, bytes32 location, uint256 price) public onlyOwner whenNotPaused {
-        require(!versions[versionHash]._exists, "This version already exists");
-        versions[versionHash] = Version({
-            id: versionHash,
-            location: location,
-            price: price,
-            _exists: true
-        });
-        latest = versionHash;
-        emit VersionCreated(sid, versionHash, location, price);
-        return;
+  function isServiceHashExist(bytes20 hash) public view returns (bool) {
+    for (uint serviceIndex = 0; serviceIndex < services.length; serviceIndex++) {
+      for (uint versionIndex = 0; versionIndex < services[serviceIndex].versions.length; versionIndex++) {
+        if (services[serviceIndex].versions[versionIndex].hash == hash) {
+          return true;
+        }
+      }
     }
+    return false;
+  }
+
+  // Getters
+
+  function getServiceVersion(uint serviceIndex, uint versionIndex) public view returns (bytes20 hash, bytes memory url) {
+    Version storage version = services[serviceIndex].versions[versionIndex];
+    return (version.hash, version.url);
+  }
+
+  function getServicePayment(uint serviceIndex, uint paymentIndex) public view returns (address purchaser) {
```

Comment:
s/getServicePayment/getServicePaymentAddress/ ?

Reply:
For now, the struct payment only contain the purchaser address. But in the future, it should have more parameters.